hal/src/rtl872x/interrupts_hal.cpp

```diff
@@ -78,48 +80,90 @@ class InterruptConfig {
     }
     ~InterruptConfig() = default;
 
+    void sortHandlersByPriority() {
+        // NOTE: has to be called under ATOMIC_BLOCK
+        std::sort(callbacks_.begin(), callbacks_.end(), [](const InterruptCallback& lhs, const InterruptCallback& rhs) {
+            // Ascending order
+            return lhs.chainPriority < rhs.chainPriority;
+        });
+    }
+
     int appendHandler(hal_interrupt_handler_t handler, void* data, uint8_t chainPriority) {
-        size_t idx = 0;
-        InterruptCallback cb = {};
+        CHECK_TRUE(handler, SYSTEM_ERROR_INVALID_ARGUMENT);
+
+        InterruptCallback cb;
         cb.handler = handler;
         cb.data = data;
         cb.chainPriority = chainPriority;
-        for (auto& callback : callbacks_) {
-            if (chainPriority < callback.chainPriority) {
-                break;
+        bool needAppend = true;
+        ATOMIC_BLOCK() {
+            for (auto& callback: callbacks_) {
+                if (!callback.handler) {
+                    callback = cb;
+                    // Used an empty entry
+                    sortHandlersByPriority();
+                    needAppend = false;
+                    break;
+                }
             }
-            idx++;
         }
-        auto temp = callbacks_;
-        CHECK_TRUE(temp.insert(idx, cb), SYSTEM_ERROR_NO_MEMORY);
-        ATOMIC_BLOCK() { 
-            spark::swap(temp, callbacks_);
+        if (!needAppend) {
+            return 0;
+        }
+
+        // Need to increase the size of the vector, which we cannot do from an ISR context
+        CHECK_FALSE(hal_interrupt_is_isr(), SYSTEM_ERROR_INVALID_STATE);
+        decltype(callbacks_) temp;
+        bool done = false;
+        while (!done) {
```

Comment:
Is the intent of the loop here to retry `Vector::reserve` / `realloc()` if it fails to allocate a new larger vector? 
Can this get stuck if we can never allocate a larger vector? I guess that is an out of memory situation and we expect to fail in other ways?

Reply:
Aaand added

Reply:
Ok so the check
```
if (temp.capacity() == (callbacks_.capacity() + 1)) {
```
is our best guess at detecting if `callbacks_` was modified after we reserved a new vector

Reply:
Correct. The modifications are either in-place in the existing buffer, or require a realloc, so capacitiy change is an indication that realloc happened, which is what we care about.

Reply:
Just to clarify realloc is always to a larger size in our vector implementation, to reduce capacity one has to manually call `trimToSize()`, this does not happen automatically. In addition to that this particular usage in interrupts_hal is also either in-place modification with InterruptCallback invalidation/replacement or append/realloc to a larger size.

Reply:
I see, when a callback is removed, its handler is set to null, but it is left in the vector. 
When a callback is added, if we find a null handler, we know it is empty and can use that item in the vector.
If we dont find a null handler, then all callbacks are valid and we need to grow `callbacks_` vector. 

For 'removed' handlers in the vector, they will all have priority `0`, so sorting by priority wont really matter for them. 
When all handlers are invoked, they will be skipped as well since they have a null callback. So no issues there.